File: Advent-of-code/2024/day-07-bridge_repair/bridge_repair.go
package main

import (
	"aoc-2024/common"
	"crypto/sha1"
	"encoding/binary"
	"fmt"
	"log"
	"math/big"
	"strconv"
	"strings"
	"unsafe"

	"github.com/mashingan/gastar"
)

const (
	input = "input.txt"
	// input = "sample.txt"
)

type node struct {
	val     uint64
	op      byte
	deep    uint
	current *big.Int
	total   *big.Int
	hash    string
}

func (n node) String() string {
	return fmt.Sprintf("{val:%d,op:%c,deep:%d,current:%d,total:%d}",
		n.val, n.op, n.deep, n.current.Uint64(), n.total.Uint64(),
	)
}

func thehash(a, b int64) string {
	h := sha1.New()
	for _, v := range []int64{a, b} {
		hs := make([]byte, unsafe.Sizeof(v))
		binary.Encode(hs, binary.NativeEndian, v)
		h.Write(hs)
	}
	s := fmt.Sprintf("%x", h.Sum(nil))
	return s
	// return fmt.Sprintf("%d|%d", a, b)
}

func (m node) Hash() string {
	return m.hash
}

type gn struct {
	gastar.Grapher[node, node, int]
	paths []*big.Int
}

func (g gn) Distance(n1, n2 node) int { return -1 }
func (g gn) Cost(n1, n2 node) int     { return 1 }

func (g gn) Neighbors(n node) []node {
	if n.deep >= uint(len(g.paths)) {
		return []node{}
	}
	currentp := big.NewInt(n.current.Int64())
	currentm := big.NewInt(n.current.Int64())
	val := n.val
	if n.deep == 0 {
		currentp = big.NewInt(g.paths[0].Int64())
		currentm = big.NewInt(g.paths[0].Int64())
	}
	if n.deep+1 < uint(len(g.paths)) {
		val = g.paths[n.deep+1].Uint64()
		currentp = currentp.Add(currentp, g.paths[n.deep+1])
		currentm = currentm.Mul(currentm, g.paths[n.deep+1])
	}
	mp := thehash(currentp.Int64(), n.total.Int64())
	mm := thehash(currentm.Int64(), n.total.Int64())
	next := []node{}
	if currentp.Cmp(n.total) <= 0 {
		next = append(next, node{
			val: val, deep: n.deep + 1, op: '+', current: currentp, total: n.total, hash: mp,
		})
	}
	if currentm.Cmp(n.total) <= 0 {
		next = append(next, node{
			val: val, deep: n.deep + 1, op: '*', current: currentm, total: n.total, hash: mm,
		})
	}
	return next
	// return []node{
	// 	{val: val, deep: n.deep + 1, op: '+', current: currentp, total: n.total, hash: mp},
	// 	{val: val, deep: n.deep + 1, op: '*', current: currentm, total: n.total, hash: mm},
	// }
}

func main() {
	log.SetFlags(log.LstdFlags | log.Lshortfile)
	tots := big.NewInt(0)
	count := 0
	validcount := 0
	common.ReadLines(input, func(text string) {
		nums := strings.Split(text, " ")
		vals := make([]*big.Int, len(nums)-1)
		sum1, _ := strconv.Atoi(strings.Trim(nums[0], ":"))
		sum := big.NewInt(int64(sum1))
		for i := 1; i < len(nums); i++ {
			n, _ := strconv.Atoi(strings.TrimSpace(nums[i]))
			vals[i-1] = big.NewInt(int64(n))
		}

		count++
		start := node{
			total:   sum,
			current: big.NewInt(0),
			hash:    thehash(0, sum.Int64()),
		}
		goal := node{
			total:   sum,
			current: sum,
			hash:    thehash(sum.Int64(), sum.Int64()),
		}
		g := gn{Grapher: gastar.NewDefault[node, node, int]()}
		g.paths = vals
		paths := gastar.PathFind[node, node](g, start, goal)
		if len(paths) <= 0 {
			return
		}
		validcount++
		tots.Add(tots, sum)
		// log.Printf("count: %d, %q\n", count, paths)
		// fmt.Printf("line: %d, the sum: %d, current sum: %d\n", count, sum.Int64(), tots.Int64())
	})
	fmt.Println(tots.Uint64())
	fmt.Println("validcount:", validcount)
}

File: Advent-of-code/2024/day-07-bridge_repair/bridge_repair.nim
#from strutils import parseInt, split
import std/[strutils, sequtils, hashes, strformat]
import bigints
from graflib import Edge, buildGraph, `a*`, `<`

type Node = object
    paths: seq[int]
    deep: int
    current: BigInt
    op: char
    total: BigInt

func hash(d: Node): Hash =
    var h: Hash = 0
    h = h !& hash(d.total)
    result = !$h

func `==`(n1, n2: Node): bool = n1.current == n2.current and n1.total == n2.total

func cost(a, b: Node): int = 1
# func distance(a, b: Node): int = 1
func distance(a, b: Node): int =
    if b.op == '+':
        if b.total - initBigInt(b.paths[b.deep]) < initBigInt(a.current): int.high
        else:
            b.deep
    else:
        let rem = b.total mod b.current
        if rem > 0'bi: int.high
        else:
            b.deep

proc next(a: Node, edges: seq[Edge[Node]]): seq[Node] =
    if a.deep + 1 > a.paths.high: return @[]
    var currentp = a.current
    var currentm = a.current
    if a.deep+1 <= a.paths.high:
        if a.deep == 0:
            currentp = initBigInt(a.paths[0] + a.paths[1])
            currentm = initBigInt(a.paths[0] * a.paths[1])
        else:
            currentp += initBigInt(a.paths[a.deep+1])
            currentm *= initBigInt(a.paths[a.deep+1])
    @[Node(
        deep: a.deep+1,
        current: currentp,
        paths: a.paths,
        total: a.total,
        op: '+',
    ), Node(
        deep: a.deep+1,
        current: currentm,
        paths: a.paths,
        total: a.total,
        op: '*',
    )]

const inputfile {.strdefine.} = "sample.txt"

proc main =
    var sum = 0'bi
    var count = 0
    var validcount = 0
    for ln in inputfile.lines:
        var g = buildGraph[Node](directed = true)
        let ll = ln.split(sep = ':', maxSplit = 2)
        let total = initBigInt(ll[0])
        # echo ll
        # echo ll[1].split.mapIt(it.strip).filterIt(it != "").map(parseInt)
        let start = Node(
            paths: ll[1].split.mapIt(it.strip).filterIt(it != "").map(parseInt),
            total: total,
        )
        let goal = Node(
            paths: start.paths,
            current: start.total,
            total: total,
        )
        var paths = `a*`[Node, int](g, start, goal)
        inc count
        if paths.len > 0:
            inc validcount
            let last = initBigInt(start.paths[^1])
            if paths[^1].current div last == paths[paths.high-1].current:
                paths[^1].op = '*'
            elif paths[^1].current - last == paths[paths.high-1].current:
                paths[^1].op = '+'
            # echo paths
            sum += total
            echo &"line: {count}, the sum: {total}, current sum: {sum}"
    echo sum
    echo "valid count: ", validcount

main()